pkg/services/searchV2/index.go

```diff
@@ -0,0 +1,359 @@
+	mu         sync.RWMutex
+	loader     dashboardLoader
+	dashboards map[int64][]dashboard
+	eventStore eventStore
+	logger     log.Logger
+}
+
+type dashboard struct {
+	id       int64
+	isFolder bool
+	folderID int64
+	slug     string
+	created  time.Time
+	updated  time.Time
+	info     *extract.DashboardInfo
+}
+
+func NewDashboardIndex(eventStore eventStore, loader dashboardLoader) *DashboardIndex {
+	return &DashboardIndex{
+		loader:     loader,
+		dashboards: map[int64][]dashboard{},
+		eventStore: eventStore,
+		logger:     log.New("dashboardIndex"),
+	}
+}
+
+func (i *DashboardIndex) listenEvents(ctx context.Context) error {
+	fullUpdateTicker := time.NewTicker(5 * time.Minute)
+	defer fullUpdateTicker.Stop()
+
+	partialUpdateTicker := time.NewTicker(5 * time.Second)
+	defer partialUpdateTicker.Stop()
+
+	var lastEventID int64
+	lastEvent, err := i.eventStore.GetLast(context.Background())
+	if err != nil {
+		return err
+	}
+	if lastEvent != nil {
+		lastEventID = lastEvent.ID
+	}
+
+	for {
+		select {
+		case <-partialUpdateTicker.C:
+			lastEventID = i.loadIndexUpdates(context.Background(), lastEventID)
+		case <-fullUpdateTicker.C:
+			i.mu.Lock()
+			i.dashboards = map[int64][]dashboard{}
+			i.mu.Unlock()
+		case <-ctx.Done():
+			return ctx.Err()
+		}
+	}
+}
+
+func (i *DashboardIndex) loadIndexUpdates(ctx context.Context, lastEventID int64) int64 {
+	events, err := i.eventStore.GetAllAfter(context.Background(), lastEventID)
+	if err != nil {
+		i.logger.Error("can't load events", "error", err)
+		return lastEventID
+	}
+	if len(events) == 0 {
+		i.logger.Info("no events since last update")
+		return lastEventID
+	}
+	for _, e := range events {
+		i.logger.Info("processing event", "event", e)
+		err := i.applyEventOnIndex(ctx, e)
+		if err != nil {
+			i.logger.Error("can't apply event", "error", err)
+			return lastEventID
+		}
+		lastEventID = e.ID
+	}
+	return lastEventID
+}
+
+func (i *DashboardIndex) applyEventOnIndex(ctx context.Context, e *EntityEvent) error {
+	if !strings.HasPrefix(e.GRN, "database/") {
+		i.logger.Info("unknown storage", "grn", e.GRN)
+		return nil
+	}
+	parts := strings.Split(strings.TrimPrefix(e.GRN, "database/"), "/")
+	if len(parts) != 3 {
+		i.logger.Error("can't parse GRN", "grn", e.GRN)
+		return nil
+	}
+	orgIDStr := parts[0]
+	kind := parts[1]
+	dashboardIDStr := parts[2]
+	if kind != "dashboards" {
+		i.logger.Error("unknown kind in GRN", "grn", e.GRN)
+		return nil
+	}
+	orgID, err := strconv.Atoi(orgIDStr)
+	if err != nil {
+		i.logger.Error("can't extract org ID", "grn", e.GRN)
+		return nil
+	}
+	dashboardID, err := strconv.Atoi(dashboardIDStr)
+	if err != nil {
+		i.logger.Error("can't extract dashboard ID", "grn", e.GRN)
+		return nil
+	}
+	return i.applyDashboardEvent(ctx, int64(orgID), int64(dashboardID), e.EventType)
+}
+
+func (i *DashboardIndex) applyDashboardEvent(ctx context.Context, orgID int64, dashboardID int64, _ EntityEventType) error {
+	i.mu.Lock()
```

Comment:
shouldn't we also cover `loadDashboards` with mutex? can't we put this on top of this fn?

```
	i.mu.Lock()
	defer i.mu.Unlock()
```